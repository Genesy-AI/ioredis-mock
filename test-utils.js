// Runs both a regular string suite and a buffer suite
export function runTwinSuite(commandName, cb) {
  const commands = [`${commandName}Buffer`, commandName]

  return commands.forEach(command =>
    command === commandName
      ? cb(command, (a, b) => a === b)
      : cb(
          command,
          (a, b) => a.equals(Buffer.from(b))
          /*
          // @TODO: detect when it happens so we can remove this
          if (Buffer.isBuffer(a)) return a.equals(Buffer.from(b))

          console.warn('a was expected to be a Buffer:', typeof a, { a, b })
          return a === b
          // */
        )
  )
}

// Util for calling describe.skip if the suite asserts Buffer outputs, which will fail in test:browser.js due to the polyfill
export const browserSafeDescribe = command =>
  process.env.IS_BROWSER && command.endsWith('Buffer')
    ? // eslint-disable-next-line no-undef
      describe.skip
    : // eslint-disable-next-line no-undef
      describe
