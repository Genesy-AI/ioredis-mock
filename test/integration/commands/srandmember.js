import Redis from 'ioredis'
import { convertBufferToString } from 'ioredis/built/utils'

// eslint-disable-next-line import/no-relative-parent-imports
import { browserSafeDescribe, runTwinSuite } from '../../../test-utils'

runTwinSuite('srandmember', command => {
  browserSafeDescribe(command)(command, () => {
    const redis = new Redis()

    afterAll(() => {
      redis.disconnect()
    })

    it('should return a random item', async () => {
      expect.assertions(1)
      await redis.sadd('myset', 'one', 'two', 'three')

      const result = convertBufferToString(await redis[command]('myset'))
      expect(['one', 'two', 'three']).toContain(result)
    })

    it('should return random unique items', async () => {
      expect.assertions(3)
      await redis.sadd('myset', 'one', 'two', 'three')

      const results = convertBufferToString(await redis[command]('myset', 2))
      expect(['one', 'two', 'three']).toContain(results[0])
      expect(['one', 'two', 'three']).toContain(results[1])
      expect(results[0]).not.toBe(results[1])
    })

    it('should return set if positive count is bigger than set', async () => {
      expect.assertions(3)
      await redis.sadd('myset', 'one', 'two', 'three')

      const results = convertBufferToString(await redis[command]('myset', 5))
      expect(['one', 'two', 'three']).toContain(results[0])
      expect(['one', 'two', 'three']).toContain(results[1])
      expect(['one', 'two', 'three']).toContain(results[2])
    })

    it('should return random items with specified length', async () => {
      expect.assertions(1)
      await redis.sadd('myset', 'one', 'two', 'three')

      const results = convertBufferToString(await redis[command]('myset', -5))
      expect(results.length).toBe(5)
    })

    it('should return null if set is empty', async () => {
      expect.assertions(1)

      const result = convertBufferToString(await redis[command]('myset'))
      expect(result).toBe(null)
    })

    it('should throw an exception if the key contains something other than a set', async () => {
      expect.assertions(1)
      await redis.set('foo', 'not a set')

      try {
        await redis[command]('foo')
      } catch (err) {
        expect(err.message).toMatchSnapshot()
      }
    })
  })
})
