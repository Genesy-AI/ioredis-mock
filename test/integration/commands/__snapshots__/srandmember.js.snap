// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`srandmember should throw an exception if the key contains something other than a set 1`] = `"WRONGTYPE Operation against a key holding the wrong kind of value"`;

exports[`srandmemberBuffer should throw an exception if the key contains something other than a set 1`] = `"WRONGTYPE Operation against a key holding the wrong kind of value"`;
